### PR TITLE
Ians Version

### PR DESCRIPTION
**Pull request by ianmann56**
I also added a commit to master of BudgetTracker Business layer: https://github.com/BudgetSquirrel/BudgetTracker/commit/da41dd261421066020f59bbb8f7b85ee2926efce
So review that commit too I accidently pushed it to master so it's already merged but review it anyways..

Also review https://github.com/BudgetSquirrel/BudgetTracker/pull/96

### CODE REVIEW

File: BudgetTracker.BudgetSquirrel.Web/Pages/Index.cshtml.cs [hunk 1]
```diff
@@ -17,34 +18,56 @@ public class IndexModel : BasePageModel
...
+                    // User has multiple or no root budgets so show the list for
+                    // them to select the one to view.
+                    AvailableRootBudgets = rootBudgets;
+                }
+            }
             (StartDate, EndDate) = GetDateWindow();
         }
 
-        public async Task<IActionResult> OnGet()
+        public async Task<IActionResult> OnGet(Guid? bid)
```
Comment (BenjaminGroseclose) [inline BudgetTracker.BudgetSquirrel.Web/Pages/Index.cshtml.cs]: `bid` isn't good for readable code. Could be something like `budgetId`

**Comment by ianmann56** [comment]
New idea... There will be a concept of income. The income object is a transaction that does not get applied until the end of the budget period (evaluated in the close method on the budget object). This way, all the income is tracked but not applied until it can be compared with what income was planned. So no transactions will be applied to the root budget, only income objects will be applied to root budget.

File: BudgetTracker.BudgetSquirrel.Web/Application/AppConstants.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+namespace BudgetTracker.BudgetSquirrel.Application
+{
+    public class AppConstants
+    {
+        public const double BUDGET_STATUS_WARNING_THRESHOLD = 0.33;
+        public const double BUDGET_STATUS_BAD_THRESHOLD = 0.05;
```
Comment (BenjaminGroseclose) [inline BudgetTracker.BudgetSquirrel.Web/Application/AppConstants.cs:6]: So if they go below 33% we give them a warning and 5% if is an another warning?

File: BudgetTracker.BudgetSquirrel.Web/Application/Auth/UserService.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        public UserService(BudgetCreation budgetCreator, IUserRepository userRepository)
+        {
+            _budgetCreator = budgetCreator;
+            _userRepository = userRepository;
+        }
+
+        public async Task<(User, Budget)> RegisterUserAndRootBudget(CreateUserViewModel setupInput)
+        {
+            EnsurePasswordConfirmation(setupInput.Password, setupInput.ConfirmPassword);
```
Comment (BenjaminGroseclose) [inline BudgetTracker.BudgetSquirrel.Web/Application/Auth/UserService.cs:23]: I usually like to `this.` for private methods so it easier to see where they came from.

File: BudgetTracker.BudgetSquirrel.Web/Application/Budgeting/BudgetCreationViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace BudgetTracker.BudgetSquirrel.Application.Budgeting
+{
+    public class BudgetCreationViewModel
+    {
+        public string Name { get; set; }
+        public double? PercentAmount { get; set; }
+        public decimal? SetAmount { get; set; }
```
Comment (BenjaminGroseclose) [inline BudgetTracker.BudgetSquirrel.Web/Application/Budgeting/BudgetCreationViewModel.cs:13]: This is probably fine for now, but might be a good idea to have a small description of the budget. Maybe the name is enough though.

File: BudgetTracker.BudgetSquirrel.Web/Pages/Components/Budget/Default.cshtml [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+@using BudgetTracker.BudgetSquirrel.Application.Budgeting
+@using BudgetTracker.BudgetSquirrel.Web.Pages
+@using BudgetTracker.BudgetSquirrel.Web.ViewComponents
+@using BudgetTracker.Business.Budgeting
+
+@model BudgetViewModel
+
+<div class="budget-item"
+        data-status="@Html.Raw((int) Model.StatusSummary)"
+        data-budget-id="@Model.Budget.Id">
```
Comment (BenjaminGroseclose) [inline BudgetTracker.BudgetSquirrel.Web/Pages/Components/Budget/Default.cshtml:11]: I am not familiar with `data-status` and `data-budget-id` how does it work.

Comment (BenjaminGroseclose) [reply BudgetTracker.BudgetSquirrel.Web/Pages/Components/Budget/Default.cshtml:11]: https://www.sitepoint.com/use-html5-data-attributes/

**Review by BenjaminGroseclose — APPROVED** [review]
Just a few small things, was a lot to look at 😄

**Comment by ianmann56** [comment]
lol @BenjaminGroseclose  this is an inactive project. You're looking for https://github.com/BudgetSquirrel/ApplicationWebApi

**Comment by BenjaminGroseclose** [comment]
Lolololol sorry

**Pull request closed without merging**